contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,781 @@
+# Scheduling Policy
+
+_Status: Draft_
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work
+-  How NodeSelector works
+-  How node / pod affinity / anti-affinity rules work
+-  How several schedulers can be used within a single cluster
+-  How priority classes work
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namepace when
+-  using schedulers,
+-  placing pods on specific nodes (master roles for instance),
+-  using specific priority classes,
+-  expressing pod affinity or anti-affinity rules.
+
+Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+Implementing SchedulingPolicy implies:
+-  Creating a new resource named **SchedulingPolicy** (schedpol)
+-  Creating an **AdmissionController** that dehaves on a deny-all-but basis
+-  Allow SchedulingPolicy to be used by pods using RoleBindings or ClusterRoleBindings
+
+# Detailed Design
+
+SchedulingPolicy resource specs are composed of several main attributes:
+-  **Required** scheduling components. These list the manadory NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations and optionally valid values that have to be provided to allow scheduling.
+-  **Allowed** scheduling components. These list the optional components that can be specified in pods definition.
```

Comment:
"Allowed" specifies the only properties that are allowed. Anything else should be rejected if specified. For example, if the only allowed "Toleration" is "label1:NoSchedule", a Pod that specified any toleration other than this should be rejected.

Reply:
In that case, is a "Forbiden" policy still necessary?

Reply:
different properties of a pod are safe in different directions... I don't think deny by default makes sense for all scheduling components

the more nodeSelector components a pod adds, the more restrictive it is. a typical policy requires some set of nodeSelector components be set on user pods (master=false, etc) but doesn't care if they specify additional selectors (worst-case, their pod cannot be scheduled)

the more tolerations a pod adds, the more permissive it is. a policy could:
* deny all tolerations by default and allow specific tolerations
* forbid specific tolerations

Reply:
Ok, got it.
So you're thinking of something like:

```yaml
# rule one: forbid a specific toleration but allow any others
specs:
  tolerations:
    deny:
    - master=true:NoSchedule
---
# rule two: allow none but a specific toleration
specs:
  tolerations:
    allow:
    - dmz=true:NoSchedule
```
With some kind of rule such as "deny and allow can't be used together on the same component"?

How do you suggest to deal with conflict resolution when multiple policies apply?

Reply:
@arnaudmz @bsalamat @liggitt - yup the way to go would be in this case, a first-seen wins ever.

Reply:
How do we expect taints and labels to be managed? It would help to have some models in mind. 

Any lists that have to be exhaustive will be hard to maintain across lots of namespaces.

Reply:
Also, I don't know what "first seen" means here.